File: dataAnalysis/languageModel/bigramReconstruct.py
import numpy as np
import os
import re
import math

modelName = 'output/bigramModels/regular.model.txt'
inputFile = 'input/2.txt'
outputFile = 'input/2.out.txt'


f = open(modelName, 'r')
lines = [ x.strip() for x in f.readlines() ]
f.close()

segToken = [ idx for idx, line in enumerate(lines) if line == '==========' ]

# unigram
idx2word = {}
word2idx = {}
unigram = {}
ta = segToken[0] + 1
tb = segToken[1]
for idx, line in enumerate( lines[ta:tb] ):
    (w, p) = line.split('\t')
    unigram[idx] = float(p)
    idx2word[idx] = w
    word2idx[w] = idx

# bigram
bigram = {}
ta = segToken[1] + 1
tb = segToken[2]
for line in lines[ta:tb]:
    (wa, wb, p) = line.split('\t')
    idxa = word2idx[wa]
    idxb = word2idx[wb]
    bigram[ (idxa, idxb) ] = float(p)

# default bigram
bigramDefault = {}
ta = segToken[2] + 1
tb = segToken[3]
for line in lines[ta:tb]:
    (w, p) = line.split('\t')
    idx = word2idx[w]
    bigramDefault[idx] = float(p)

# read input
f = open(inputFile, 'r')
content = f.read()
f.close()


nrWords = len(unigram) + 1   # last word as empty
lenContent = len(content)
dummyWordIdx = nrWords - 1

nonMatchPenalty = math.log10( 1.0 / nrWords )
ninf = -1000000000.0

dp = np.empty( [ (lenContent+1), nrWords ] )
dp.fill(ninf)
fromCIdx = np.empty( [ (lenContent+1), nrWords ] )  # character index
fromWIdx = np.empty( [ (lenContent+1), nrWords ] )  # word index

dp[0][dummyWordIdx] = 0.0

for ncidx in range(lenContent):
    print('ncidx: ', ncidx, ' /', lenContent)
    for nwidx in range(nrWords):
        if dp[ncidx][nwidx] == ninf:
            continue
        
        # option 1: go for penalty
        tcidx = ncidx + 1
        twidx = nwidx
        if dp[ncidx][nwidx] + nonMatchPenalty > dp[tcidx][twidx]:
            dp[tcidx][twidx] = dp[ncidx][nwidx] + nonMatchPenalty
            fromCIdx[tcidx][twidx] = ncidx
            fromWIdx[tcidx][twidx] = nwidx

        for twidx in range(nrWords-1):
            tcidx = ncidx + len( idx2word[twidx] )
            #print(tcidx, '/', lenContent, twidx, '/', nrWords)
            if tcidx <= lenContent and content[ncidx:tcidx] == idx2word[twidx]:
                tvalue = dp[ncidx][nwidx] + unigram[twidx]
                if tvalue > dp[tcidx][twidx]:
                    dp[tcidx][twidx] = tvalue
                    fromCIdx[tcidx][twidx] = ncidx
                    fromWIdx[tcidx][twidx] = nwidx
                    
endcIdx = lenContent - 1
endwIdx = 0
for widx in range(nrWords-1):
    if dp[endcIdx][widx] > dp[endcIdx][endwIdx]:
        endwIdx = widx

ansWords = []
while endcIdx > 0:
    #print('==>', endcIdx, endwIdx, idx2word[endwIdx])
    ansWords += [ idx2word[endwIdx] ]
    pcidx, pwidx = fromCIdx[endcIdx][endwIdx], fromWIdx[endcIdx][endwIdx]
    endcIdx, endwIdx = pcidx, pwidx

ansWords.reverse()
#print(ansWords)
f = open(outputFile, 'w')
f.write( ' '.join( tuple(ansWords) ) )
f.close()

File: dataAnalysis/languageModel/bigramReconstruct.cpp
#include <stdio.h>
#include <vector>
#include <map>
#include <string>
#include <string.h>
#include <math.h>

#include "bigramModel.cpp"

using namespace std;

const char *modelName = "output/bigramModels/regular.model.txt";
const char *inputFile = "input/7.txt";
const char *outputFile = "input/7.out.txt";

const int MAX_INPUT_SIZE = 50000;

const double NINF = -10000000.0;

template <class T>
T** make2dArray(int na, int nb) {
	T** re = new T*[na];
	for (int i = 0; i < na; i++)
		re[i] = new T[nb];
	return re;
}


double charReplacementProb(const char *wa, const int lena, const char *wb, const int lenb) {
	if (lena != lenb)
		return NINF;
	for (int i = 0; i < lena; i++)
		if (wa[i] != wb[i])
			return NINF;
	return 0.0;
}


int main() {
	BigramModel model(modelName);

	printf("nrWords = %d\n", model.nrWords);

	char content[MAX_INPUT_SIZE];
	FILE *fin = fopen(inputFile, "r");
	fread(content, sizeof(char), MAX_INPUT_SIZE, fin);
	fclose(fin);

	int contentLen = strlen(content);

	printf("content length = %d\n", contentLen);

	const double penalityOfNonMatch = log10( 1.0 / model.nrWords / 2.0 );

	double **dp = make2dArray<double>(contentLen+1, model.nrWords);
	int **fromCIdx = make2dArray<int>(contentLen+1, model.nrWords);
	int **fromWIdx = make2dArray<int>(contentLen+1, model.nrWords);

	for (int i = 0; i < contentLen+1; i++)
		for (int j = 0; j < model.nrWords; j++)
			dp[i][j] = NINF;

	dp[0][ model.widxDummy ] = 0.0;

	int tcidx, twidx;
	double tvalue;
	for (int ncidx = 0; ncidx < contentLen; ncidx++) {
		printf("ncidx=%d (/%d)\n", ncidx, contentLen);
		for (int nwidx = 0; nwidx < model.nrWords; nwidx++) {
			if (dp[ncidx][nwidx] == NINF)
				continue;

			// option 1: go for penalty
			tcidx = ncidx + 1;
			twidx = model.widxDummy;
			tvalue = dp[ncidx][nwidx] + penalityOfNonMatch;
			if (tvalue > dp[tcidx][twidx]) {
				dp[tcidx][twidx] = dp[ncidx][nwidx] + tvalue;
				fromCIdx[tcidx][twidx] = ncidx;
				fromWIdx[tcidx][twidx] = nwidx;
			}

			// option 2: go for bigram
			for (twidx = 0; twidx < model.nrWords-1; twidx++) {
				tcidx = ncidx + model.wordLen[twidx];
				if (tcidx <= contentLen) {
					double replaceProb = charReplacementProb(
							content+ncidx, model.wordLen[twidx],
							model.idx2word[twidx].c_str(), model.wordLen[twidx]);
					//for (int i = 0; i < ncidx; i++)
					//	printf("%c", content[ncidx + i]);
					//printf(" %s %lf\n", model.idx2word[twidx].c_str(), replaceProb);
					double tvalue = dp[ncidx][nwidx] + model.unigram[twidx] + replaceProb;
					if (tvalue > dp[tcidx][twidx]) {
						dp[tcidx][twidx] = tvalue;
						fromCIdx[tcidx][twidx] = ncidx;
						fromWIdx[tcidx][twidx] = nwidx;
					}
				}
			}
		}
	}

	vector<string> ans;

	int endCIdx = contentLen - 1;
	int endWIdx = 0;
	for (int widx = 1; widx < model.nrWords-1; widx++)
		if (dp[endCIdx][widx] > dp[endCIdx][endWIdx])
			endWIdx = widx;


	while (endCIdx > 0) {
		if (endWIdx == model.widxDummy)
			ans.push_back("?");
		else
			ans.push_back( model.idx2word[endWIdx] );

		int pcidx = fromCIdx[endCIdx][endWIdx];
		int pwidx = fromWIdx[endCIdx][endWIdx];
		endCIdx = pcidx;
		endWIdx = pwidx;
	}

	FILE *fout = fopen(outputFile, "w");
	for (int i = ans.size() - 1; i >= 0; i--)
		fprintf(fout, "%s ", ans[i].c_str());
	
	fclose(fout);

	// lazy delete... I hope C++ can do GC...

	return 0;
}
